Develop upload clan inventory

Soda-xyz opened the pull request:
Vibe: pleased
Possibly a lot of unclear naming.
Feel like there might be some readability issues due to size.
Mainly
upload_clan_inventory.py
and 
assign_hibernation.py

Soda-xyz committed:
refactor: Updated table structure

Soda-xyz committed:
feat: slash command to upload clan inventory

Soda-xyz committed:
feat: slash command to upload clan inventory

Soda-xyz committed:
feat: crud to upsert warframe_player table.

Soda-xyz committed:
feat: crud to upsert warframe_player table.

Soda-xyz committed:
Merge remote-tracking branch 'origin/develop-upload-clan-inventory' into develop-upload-clan-inventory

# Conflicts:
#	obearon/discord/commands/upload_clan_inventory.py

Soda-xyz committed:
refactor: adherence to normalization for clan inventory data

Soda-xyz committed:
refactor: upload slash command for new db structure

Soda-xyz committed:
refactor: crud commands for upsert clan inventory and retrieve names.

Soda-xyz committed:
refactor: assign_hibernation to event. Moved to events folder

Soda-xyz committed:
refactor: hibernation assignment reworked to auto trigger post upload

Soda-xyz committed:
Merge branch 'main' into develop-upload-clan-inventory

akshualy commented on obearon/database/crud/warframe_players.py:
```diff
@@ -0,0 +1,89 @@
+from obearon.database import models
+
+
+async def create_update_warframe_players(players: list[dict], player_names: list[dict]):
```
A few things:
1. I would prefer if these were not untyped dictionaries.
2. I would prefer if these were not dictionaries at all. Like, *we* know how the data looks like, but I'd prefer access to an object that IDEs can work with instead of access by arbitrary name to a dictionary. Feel free to use pydantic or dataclasses if you agree.
3. Why are we distinguishing players and player_names? Even if we distinguish them in the database, we should not distinguish them in the data flow.

Soda-xyz replied:
I agree, not using dictionaries seems like a good idea.

3. Can you elaborate on this?
In my mind I am distinguishing them to have a collection of **players** that are for the **warframe_players** table and the relevant info for that. And the **player_names** for the **warframe_player_names** table.
Is this approach needlessly segmented?

akshualy replied:
If `players` is `list[Player]`, you would do `player_names = [player.names for player in players]`.
Idk, I think this unnecessarily separates names from players and I would prefer us reflecting the structure DE has.

akshualy commented on obearon/database/crud/warframe_players.py:
```diff
@@ -0,0 +1,89 @@
+            )
+            await session.execute(upsert)
+        out_of_clan = (
+            update(models.WarframePlayers).where(models.WarframePlayers.oid.not_in(player_oids)).values(in_clan=False)
```
I think you should just brute update all fields for in_clan to False by dropping this where clause. And move the update to before the upsert. `on_conflict_do_update` will fire for all entries anyway. `not in` is really slow.

akshualy commented on obearon/database/crud/warframe_players.py:
```diff
@@ -0,0 +1,89 @@
+                    index_elements=["oid"], set_={"mastery_rank": player["mastery_rank"], "in_clan": True}
+                )
+            )
+            await session.execute(upsert)
```
This is a note and you can just resolve it without comment; this is only okay because we're using transactions. When doing many SQL updates, you usually batch them first.

Soda-xyz replied:
I will make an issue cause I want to get back to it and figure it out later. Is it correct that 'bulk_"x"_mappings' I should be looking at for this?

akshualy replied:
https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-enabled-insert-update-and-delete-statements

but as i said, this doesn't matter as much, since a transaction bundles the inserts into one update anyway

akshualy commented on obearon/database/crud/warframe_players.py:
```diff
@@ -0,0 +1,89 @@
+
+    async with engine.async_session() as session, session.begin():
+        names = await session.execute(select(models.WarframePlayerNames.name))
+        return [row[0] for row in names.all()]
```
Since you select only one column and that column is a string, using `return names.scalars().all()` will already return a `list[str]`. No need to convert/unwrap. `scalars` will do it for you.

akshualy commented on obearon/database/crud/warframe_players.py:
```diff
@@ -0,0 +1,89 @@
+    )
+    async with engine.async_session() as session, session.begin():
+        names = await session.execute(fetch_names)
+        return [row[0] for row in names.all()]
```
Same here about scalars.

akshualy commented on obearon/database/models/warframe_player_names.py:
```diff
@@ -0,0 +1,25 @@
+    """
+
+    __tablename__ = "warframe_player_names"
+    __table_args__ = (UniqueConstraint("oid", "name", "platform"),)
```
You should give constraints a name. For example, `name="uq_oid_name_platform"`.

Soda-xyz replied:
What is the name used for? Error logging for the database?

akshualy commented on obearon/database/models/warframe_player_names.py:
```diff
@@ -0,0 +1,25 @@
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, unique=True, autoincrement=True, nullable=False)
+    oid: Mapped[str] = mapped_column(String, ForeignKey("warframe_players.oid"), nullable=False)
+    name: Mapped[str] = mapped_column(String, nullable=False)
+    platform: Mapped[str] = mapped_column(String, nullable=False, default="unknown")
```
Can this ever be "unknown"? `default` is client-side and not server-side by the way.

akshualy commented on obearon/database/models/warframe_player_names.py:
This file is missing a migration.

akshualy commented on obearon/discord/commands/hibernation_role.py:
```diff
@@ -47,6 +47,7 @@ async def get_hibernation_role(self, ctx: discord.ApplicationContext) -> None:
         Args:
             ctx: The context of the user who executed this command.
         """
+
```
? You should really get into the habit of using the linter tools :D I'll make sure to add pre-commit soon.

Soda-xyz replied:
I try ;_;
I gave a decent effort at having IDE force me to lint before commit, but the plugins haven't been ported to the version I use or something like that.

akshualy commented on obearon/discord/commands/upload_clan_inventory.py:
```diff
@@ -0,0 +1,104 @@
+            "\ue002": "playstation",
+            "\ue003": "nintendo",
+            "\ue004": "ios",
+            "\ue005": "unk005",
+            "\ue006": "unk006",
+            "\ue007": "unk007",
```
Let's not future-proof and just update when we know what they are.

akshualy commented on obearon/discord/commands/upload_clan_inventory.py:
```diff
@@ -0,0 +1,104 @@
+                all_names = [member["DisplayName"]]
+                all_names.extend(member.get("PlatformNames", []))
+                for name in all_names:
+                    match = re.match(r"(.*?)([\ue000-\uefff])$", name)
```
Why match on such a wide range if we know what's valid?
It's also always only one character, so no need to use an uncompiled regex and string slice instead.

akshualy commented on obearon/discord/commands/upload_clan_inventory.py:
```diff
@@ -0,0 +1,104 @@
+            file_content = await file.read()
+            bytes_to_string = file_content.decode("utf-8")
+            parsed_json = json.loads(bytes_to_string)
+            for member in parsed_json["Members"]:
```
Same comment here about dumping into models before accessing things.

akshualy commented on obearon/discord/commands/upload_clan_inventory.py:
```diff
@@ -0,0 +1,104 @@
+                    if match:
+                        player_name = match.group(1)
+                        platform_code = match.group(2)
+                        platform = platform_dict.get(platform_code, "unknown")
```
Honestly, I really prefer raising or logging an exception over having "unknown" in the database.

akshualy commented on obearon/discord/events/assign_hibernation.py:
```diff
@@ -0,0 +1,45 @@
+from obearon.database import crud
+
+
+async def on_application_command_completion(ctx) -> None:
```
I honestly think this is not where I'd want this to be. Can you not just call a method after the context response?

Soda-xyz replied:
Can you elaborate on this, I am not following.

akshualy replied:
Like, why are we listening to a global event for all command executions if we could just run this code after the command executes inside of the command itself?

akshualy commented on obearon/discord/events/assign_hibernation.py:
```diff
@@ -0,0 +1,45 @@
+            joined_recently = datetime.now(timezone.utc) - member.joined_at < timedelta(hours=24)
+            if joined_recently:
+                continue
+            if not any(name.lower() == member.nick.lower() for name in warframe_players):
```
Another note for you that you can just ignore if you want to, `.lower()` for matching strings is only okay because we know that warframe names are `[a-z0-9_.-]`. For more complex cases, you'd want `casefold()`.

Soda-xyz committed:
refactor: removed default value from in_clan.
removed "future proofing" platform names.
in_clan=False now brute forced before new data.
+ Migration file for player tables

Soda-xyz committed:
refactor: migration file to include constraint name. Upload now compares unicode and defaults to pc
